File: Diferenca_Finita/plot.py
#!/usr/bin/python

import numpy as np
import matplotlib.pyplot as plot
import sys

data = np.genfromtxt("solution.dat",delimiter=' ',names=['x','analit','aprox','residue'])

# Grafico da solucao analitica vs aproximacao
fig = plot.figure()
axis = fig.add_subplot(111)
axis.set_title("Diferencas Finitas")
axis.set_xlabel('x')
axis.set_ylabel('y')
axis.plot(data['x'],data['analit'],c='r',label='analit')
axis.plot(data['x'],data['aprox'],c='b',label='aprox')
leg = axis.legend(loc=2)
plot.grid()
#plot.savefig('/home/lucas/Documentos/Mestrado/Metodo_Elementos_Finitos/Projecao_L2/solucao.png')
plot.show()

# Grafico do residuo
fig = plot.figure()
axis = fig.add_subplot(111)
axis.set_title("Diferencas Finitas - Residuo")
axis.set_xlabel('x')
axis.set_ylabel('R')
axis.plot(data['x'],data['residue'],c='g',label='residue')
leg = axis.legend(loc=2)
plot.grid()
#plot.savefig('/home/lucas/Documentos/Mestrado/Metodo_Elementos_Finitos/Projecao_L2/residuo.png')
plot.show()

File: Diferenca_Finita/src/plot.c
#include "plot.h"

double Faprox (double x, Func *f, double *alfa, int N)
{
  int k;
  double eval = 0;
  for (k = 0; k < N; k++)
    eval += alfa[k]*f[0](k,x);
  return eval;
}

double Fanalit (double x, double a, double b, double c)
{
  return ((exp(b*x/a)-1) / (exp(b/a)-1));
}

void writeDataFile (double *u, double a, double b, double c, double h, int N)
{
  FILE *file_solution, *file_residue;
  int i, n;
  double x, fanalit, faprox, residue;

  file_solution = fopen("solution.dat","w+");
  file_residue = fopen("residue.dat","w+");
  // Escreve a aproximacao
  for (i = 0; i < N; i++)
  {
    x = i*h;
    fanalit = Fanalit(x,a,b,c);
    faprox = u[i];
    residue = fabs(fanalit-faprox);
    // || x || analitica || aproximacao || residuo ||
    fprintf(file_solution,"%e %e %e %e\n",x,fanalit,faprox,residue);
    fprintf(file_residue,"%e %e\n",x,residue);
  }
  fclose(file_solution);
  fclose(file_residue);
}

void printMatrix (char *str, double *A, int N)
{
  int i, j;
  printf("\n%s\n",str);
  for (i = 0; i < N; i++)
  {
    printf("\n");
    for (j = 0; j < N; j++)
    {
      printf("%e ",A[i * N + j]);
    }
  }
  printf("\n");
}

void printVector (char *str, double *b, int N)
{
  int i, j;
  printf("\n%s\n",str);
  for (i = 0; i < N; i++)
    printf("%e\n",b[i]);
  printf("\n");
}
